### ISSUE TITLE
Unable to use Keras data generator with Functional API model

### ISSUE THREAD

**Issue by Adgoswa** [tensorflow/tensorflow#50745]
Here is my data generator class:
class DataGenerator(tensorflow.keras.utils.Sequence):
    'Generates data for Keras'
    def __init__(self, list_IDs, image_path, label_path,
                 to_fit=True, batch_size=16, dim=(160, 160),
                 n_channels=4, n_classes=4, shuffle=True):
        """Initialization
        :param list_IDs: list of all 'label' ids to use in the generator
        :param labels: list of image labels (file names)
        :param image_path: path to images location
        :param mask_path: path to masks location
        :param to_fit: True to return X and y, False to return X only
        :param batch_size: batch size at each iteration
        :param dim: tuple indicating image dimension
        :param n_channels: number of image channels
        :param n_classes: number of output masks
        :param shuffle: True to shuffle label indexes after every epoch
        """
        self.list_IDs = list_IDs
        #self.labels = labels
        self.image_path = image_path
        self.label_path = label_path
        self.to_fit = to_fit
        self.batch_size = batch_size
        self.dim = dim
        self.n_channels = n_channels
        self.n_classes = n_classes
        self.shuffle = shuffle
        self.on_epoch_end()
    
    def __len__(self):
        'Denotes the number of batches per epoch'
        #print('Batch per epoch: ', self.batch_per_epoch)
        #return self.batch_per_epoch
        return int(np.floor(len(self.list_IDs) / self.batch_size))

    def on_epoch_end(self):
        'Updates indexes after each epoch'
        self.indexes = np.arange(len(self.list_IDs))
        if self.shuffle == True:
            np.random.shuffle(self.indexes)

    def __getitem__(self, index):
        'Generate one batch of data'
        #data = np.load('.\\SplitDataSet\\Data\\x_1.npy')
        #labels = np.load('.\\SplitDataSet\\Label\\y_1.npy')
        """list_IDs_temp = []
        for i in range(self.start_index, self.end_index + 1):
            list_IDs_temp.append(i)"""
        # Generate indexes of the batch
        indexes = self.indexes[index * self.batch_size:(index + 1) * self.batch_size]

        # Find list of IDs
        list_IDs_temp = [self.list_IDs[k] for k in indexes]
        #print(list_IDs_temp)
        X = self._generate_X(list_IDs_temp)
        print('X shape', X.shape)
        y = self._generate_y(list_IDs_temp)
        print('Y shape', y.shape)
        return X, y

    def _generate_X(self, list_IDs_temp):
        # Initialization
        X = np.empty((self.batch_size, 155, *self.dim, self.n_channels))
        for i, ID in enumerate(list_IDs_temp):
            X[i,] = np.load('./SplitDataSet/Data_4/x_' + str(ID) + '.npy')
            #test = np.load('.\\SplitDataSet\\Data\\x_' + str(ID) + '.npy')
            #print('File name: ', './SplitDataSet/Data/x_' + str(ID) + '.npy')
            #print(test.shape)
        X = X.reshape([-1,160,160,4])
        return X
    
    def _generate_y(self, list_IDs_temp):
        # Initialization
        y = np.empty((self.batch_size, 155, *self.dim, self.n_channels))
        #y = np.empty((self.batch_size, 155, *self.dim, 2))
        for i, ID in enumerate(list_IDs_temp):
            y[i,] = np.load('./SplitDataSet/Label_4/y_' + str(ID) + '.npy')
        y = y.reshape([-1,160,160,4])
        return y

And my model looks like this:
input_ = Input(shape=(160, 160, 4), name='input')

#   EXAMPLE ENCODER
block1_conv1 = Conv2D(64, 3, padding='same', activation='relu', name='block1_conv1')(input_)
block1_conv2 = Conv2D(64, 3, padding='same', activation='relu', name='block1_conv2')(block1_conv1)
block1_norm = BatchNormalization(name='block1_batch_norm')(block1_conv2)
block1_pool = MaxPooling2D(name='block1_pool')(block1_norm)

block2_conv1 = Conv2D(128, 3, padding='same', activation='relu', name='block2_conv1')(block1_pool)
block2_conv2 = Conv2D(128, 3, padding='same', activation='relu', name='block2_conv2')(block2_conv1)
block2_norm = BatchNormalization(name='block2_batch_norm')(block2_conv2)
block2_pool = MaxPooling2D(name='block2_pool')(block2_norm)

dropout_1 = Dropout(0.2, name='encoder_dropout_1')(block2_pool)

block3_conv1 = Conv2D(256, 3, padding='same', activation='relu', name='block3_conv1')(dropout_1)
block3_conv2 = Conv2D(256, 3, padding='same', activation='relu', name='block3_conv2')(block3_conv1)
block3_norm = BatchNormalization(name='block3_batch_norm')(block3_conv2)
block3_pool = MaxPooling2D(name='block3_pool')(block3_norm)

block4_conv1 = Conv2D(512, 3, padding='same', activation='relu', name='block4_conv1')(block3_pool)
block4_conv2 = Conv2D(512, 3, padding='same', activation='relu', name='block4_conv2')(block4_conv1)
block4_norm = BatchNormalization(name='block4_batch_norm')(block4_conv2)
block4_pool = MaxPooling2D(name='block4_pool')(block4_norm)

#BOTTOM OF U
block5_conv1 = Conv2D(1024, 3, padding='same', activation='relu', name='block5_conv1')(block4_pool)

#   EXAMPLE DECODER

up_pool1 = Conv2DTranspose(1024, 3, strides=(2, 2), padding='same', activation='relu', name='up_pool1')(block5_conv1)
merged_block1 = concatenate([block4_norm, up_pool1], name='merged_block1')
decod_block1_conv1 = Conv2D(512, 3, padding='same', activation='relu', name='decod_block1_conv1')(merged_block1)

up_pool2 = Conv2DTranspose(512, 3, strides=(2, 2), padding='same', activation='relu', name='up_pool2')(decod_block1_conv1)
merged_block2 = concatenate([block3_norm, up_pool2], name='merged_block2')
decod_block2_conv1 = Conv2D(256, 3, padding='same', activation='relu', name='decod_block2_conv1')(merged_block2)

dropout_2 = Dropout(0.2, name='decoder_dropout_1')(decod_block2_conv1)

up_pool3 = Conv2DTranspose(256, 3, strides=(2, 2), padding='same', activation='relu', name='up_pool3')(dropout_2)
merged_block3 = concatenate([block2_norm, up_pool3], name='merged_block3')
decod_block3_conv1 = Conv2D(128, 3, padding='same', activation='relu', name='decod_block3_conv1')(merged_block3)

up_pool4 = Conv2DTranspose(128, 3, strides=(2, 2), padding='same', activation='relu', name='up_pool4')(decod_block3_conv1)
merged_block4 = concatenate([block1_norm, up_pool4], name='merged_block4')
decod_block4_conv1 = Conv2D(64, 3, padding='same', activation='relu', name='decod_block4_conv1')(merged_block4)

pre_output = Conv2D(64, 1, padding='same', activation='relu', name='pre_output')(decod_block4_conv1)
output = Conv2D(4, 1, padding='same', activation='softmax', name='output')(pre_output)
modelUNet = Model(inputs=input_, outputs=output)
print(modelUNet.summary())

image_path = 'path to images'
label_path = 'path to masks'
total_records = 150
train_start = 0
train_end = int(np.floor(0.8*total_records))
val_start = train_end + 1
val_end = total_records
train_idx = []
val_idx = []
for i in range(1, val_start):
    train_idx.append(i)
for i in range(val_start, total_records + 1):
    val_idx.append(i)

training_generator = DataGenerator(train_idx, image_path, label_path)
validation_generator = DataGenerator(val_idx, image_path, label_path)

# The model is compiled with the dice_loss_function and the dice_function metric:
print("About to compile...")
#modelUNet = Sequential()
modelUNet.compile(optimizer=Adam(lr=1e-5), loss=dice_loss_function, metrics=[dice_function])

# EarlyStopping is applied incase the model stops improving with each epoch:
callbacks = [tf.keras.callbacks.EarlyStopping(patience=2, monitor='val_loss')]

print("About to fit the model...")

history = modelUNet.fit(
	x=training_generator,
	validation_data=validation_generator,
	epochs=10, callbacks=callbacks)

Now the issue is my model is not sequential, so when I comment out the line: modelUNet = Sequential() and run the code, it throws the following error: W tensorflow/core/kernels/data/generator_dataset_op.cc:107] Error occurred when finalizing GeneratorDataset iterator: Failed precondition: Python interpreter state is not initialized. The process may be terminated.
         [[{{node PyFunc}}]]
And the process terminates, but if I uncomment the modelUNet = Sequential(), then the model runs but since it is not sequential, it does not learn and finishes at 5 epochs because it crosses the patience.
How do I use a data generator with a non-sequential model?

**Comment by UsharaniPagadala** [comment]
@Adgoswa 

Please post this issue on [keras-team/keras](https://github.com/keras-team/keras/issues) repo.
To know more refer to:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999





 and also refer similar [issue](https://discuss.tensorflow.org/t/error-occurred-when-finalizing-generatordataset-iterator/2080
) , [link2](https://github.com/tensorflow/tensorflow/issues/35100)
